luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/rotator.rs

```diff
@@ -0,0 +1,142 @@
+            .set_anti_alias(true);
+
+        let rendering_tree = render([
+            namui::path(handle_path.clone(), fill_paint),
+            namui::path(handle_path, stroke_paint.clone()),
+            namui::path(bridge_path, stroke_paint),
+        ])
+        .with_mouse_cursor(MouseCursor::Crosshair)
+        .attach_event(move |event| match dragging_context {
+            Some(context) => match event {
+                namui::Event::MouseMove { event } => {
+                    on_event(Event::OnUpdateDraggingContext {
+                        context: Some(RotatorDraggingContext {
+                            origin_global_xy: context.origin_global_xy,
+                            end_global_xy: event.global_xy,
+                        }),
+                    });
+                }
+                namui::Event::MouseUp { event: _event } => {
+                    on_event(Event::OnUpdateDraggingContext { context: None });
+                    on_event(Event::OnRotate {
+                        rotation: context.rotation(keyboard::shift_press()),
+                        graphic_index,
+                    });
+                }
+                _ => {}
+            },
+            None => {
+                if let namui::Event::MouseDown { event } = event {
+                    if event.is_local_xy_in() && event.button == Some(MouseButton::Left) {
+                        event.stop_propagation();
+                        on_event(Event::OnUpdateDraggingContext {
+                            context: Some(RotatorDraggingContext {
+                                origin_global_xy: ctx.global_xy(rect.center()),
+                                end_global_xy: event.global_xy,
+                            }),
+                        });
+                    }
+                }
+            }
+        });
+
+        ctx.component(rendering_tree);
+
+        ctx.done()
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct RotatorDraggingContext {
+    origin_global_xy: Xy<Px>,
+    end_global_xy: Xy<Px>,
+}
+
+impl RotatorDraggingContext {
+    pub fn rotation(&self, snap: bool) -> Angle {
+        const SNAP_ANGLE_DEGREE: f32 = 15.0;
+        let angle = calculate_angle(
+            Xy::new(0.0, -1.0),
+            self.end_global_xy - self.origin_global_xy,
+        );
+        if !snap {
+            return angle;
+        }
+        Angle::Degree((angle.as_degrees() / SNAP_ANGLE_DEGREE).round() * SNAP_ANGLE_DEGREE)
+    }
+}
+
+fn calculate_angle<IntoF32A, IntoF32B>(
+    vector_2d_a: Xy<IntoF32A>,
+    vector_2d_b: Xy<IntoF32B>,
+) -> Angle
+where
+    IntoF32A: Into<f32> + Sized,
+    IntoF32B: Into<f32> + Sized,
+{
+    let a_x = vector_2d_a.x.into();
+    let a_y = vector_2d_a.y.into();
+    let b_x = vector_2d_b.x.into();
+    let b_y = vector_2d_b.y.into();
+
+    let dot = a_x * b_x + a_y * b_y;
+    let determinant = a_x * b_y - a_y * b_x;
+    Angle::Radian(determinant.atan2(dot))
+}
```

Comment:
https://github.com/NamseEnt/namseent/blob/a5c4ac2923c491fa6481c87c45cc35f934750cf4/namui-type/src/types/units/vector_types/xy.rs#L9

There are this kind of things. use that.

Reply:
Thank you!
Fixedd with 2ce763b